Fix product list refreshing

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
### Fix product list refreshing

**Closes**: #8833 
<!-- Id number of the GitHub issue this PR addresses. -->
**Problem**: In case there are many draft products (20+), e.g. when setting up a new store, and we fetch them into the local database - for example by opening the product selector screen, when we add a published product it's not visible immediately.

### Description
Re-fetching products in case no products are found in local storage.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Update all the products in the store to `status: draft` and open the product selector. Observe that no products are displayed.
2. Update the status of one product to `published`. Observe it gets fetched and displayed in the product selector.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Fix product list refreshing

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModelTest.kt:
```diff
@@ -402,6 +402,31 @@ internal class ProductSelectorViewModelTest : BaseUnitTest() {
     }
 
+    @Test
+    fun `given no published products in cache, when view model created, then should fetch products`() = testBlocking {
```
❓ I think that this test passes because of the `fetchProducts(forceRefresh = true)` in the init block of the view model. Are we sure that this test actually tests the `fetchProducts(forceRefresh = true)` from the `onEmpty` block?

samiuelson replied:
Yes, good point. It turns out not easy to unit test this edge case using mocked dependencies. I'm removing the method for now.

AnirudhBhat commented:
@jostnes @samiuelson I'm a little confused about testing this PR. 

>Update the status of one product to published. Observe it gets fetched and displayed in the product selector.

Won't this already work in the app? 

I'll explain my understanding of the problem. Please correct me if my understanding is not correct.

When there are many draft/auto-draft products (ideally more than 1 page), the product selector screen won't list even the published products as well. Is my understanding correct?

@jostnes Is there any testing store that you could invite me to so that I can test this functionality?

samiuelson commented:
> @jostnes @samiuelson I'm a little confused about testing this PR.
> 
> > Update the status of one product to published. Observe it gets fetched and displayed in the product selector.
> 
> Won't this already work in the app?
> 
> I'll explain my understanding of the problem. Please correct me if my understanding is not correct.
> 
> When there are many draft/auto-draft products (ideally more than 1 page), the product selector screen won't list even the published products as well. Is my understanding correct?
> 
> @jostnes Is there any testing store that you could invite me to so that I can test this functionality?

@AnirudhBhat I've invited you to my store (samurbanowicz3.wpcomstaging.com) which currently has multiple pages of draft products only. 

**The way I tested it:**
1. Opened the product selector and observed no products were displayed.
2. Closed the product selector
3. Updated one of the products to the `published` state
4. Reopened the product selector screen

Observations: on `trunk` no products were displayed at point 4. On the `8833-product-selector-list-refresh-improvement` branch the published product was displayed at point 4.

**My understanding:** When draft products are filtered out from the `products` flow here:
```kotlin
private val products = listHandler.productsFlow.map { products ->
        products.filter { product ->
            productsRestrictions.map { restriction -> restriction(product) }.fold(true) { acc, result -> acc && result }
        }
    }
```
In the case there are no items in the flow after the filtering operation we didn't fetch the next page automatically. BTW, we would trigger fetching it if we tried to scroll down on the empty state screen, and the missing published product would be shown. This fix is supposed to trigger fetching the next page in case there are no products left after the filtering operation. 

Edit: I moved the onEmpty {} block to the products flow directly - it fits better to the case described above.

samiuelson committed:
Move onEmpty block to products flow

jostnes commented:
> @jostnes Is there any testing store that you could invite me to so that I can test this functionality?

@AnirudhBhat sorry for the late response, I missed this ping earlier but it looks like @samiuelson has added you to a store that has the issue, let me know if you still need me to add you to mine. Thanks!

AnirudhBhat commented:
@samiuelson 
>The way I tested it:
Opened the product selector and observed no products were displayed.
Closed the product selector
Updated one of the products to the published state
Reopened the product selector screen

It looks like it works the same on `trunk`. I followed the same steps on `trunk` and was able to see the published product in step 4. I used your store (samurbanowicz3.wpcomstaging.com) for testing this on `trunk`. Not sure if I'm doing something wrong.

samiuelson commented:
Looks like the bug is no longer reproducible on the trunk.
